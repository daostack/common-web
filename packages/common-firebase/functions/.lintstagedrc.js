module.exports = {
  '**/*.js?(x)': (filenames) => `eslint ${filenames.join(' ')}`
}
