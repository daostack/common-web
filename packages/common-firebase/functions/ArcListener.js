const { Arc } = require('@daostack/arc.js');
const admin = require('firebase-admin');

const graphHttpLink = 'https://api.thegraph.com/subgraphs/name/daostack/v8_1_exp_xdai'
const graphwsLink = 'wss://api.thegraph.com/subgraphs/name/daostack/v8_1_exp_xdai'

const arc = new Arc({
  graphqlHttpProvider: graphHttpLink,
  graphqlWsProvider: graphwsLink,
});

const db = admin.firestore();

// get all DAOs data from graphql and read it into the subgraph
async function updateDaos() {
  const response = [];
  const daos = await arc.daos().first();
  console.log(`Found ${daos.length} DAOs`);
  for (const dao of daos) {
    const joinAndQuitPlugins = await dao.plugins({where: {name: 'JoinAndQuit'}}).first();
    if (joinAndQuitPlugins.length === 0) {
      // not a properly configured common DAO, skipping
      const msg = `Skipping ${dao.id} as it is not properly configured`;
      console.log(msg);
      response.push(msg)
    } else {
      console.log(`updating ${dao.id}`);
      const joinAndQuitPlugin = joinAndQuitPlugins[0];
      const {
        fundingGoal,
        minFeeToJoin,
        memberReputation
      } = joinAndQuitPlugin.coreState.pluginParams;
      try {
      const daoState = dao.coreState
      const doc = {
        id: dao.id,
        address: daoState.address,
        memberCount: daoState.memberCount,
        name: daoState.name,
        numberOfBoostedProposals: daoState.numberOfBoostedProposals,
        numberOfPreBoostedProposals: daoState.numberOfPreBoostedProposals,
        numberOfQueuedProposals: daoState.numberOfQueuedProposals,
        register: daoState.register,
        // reputationId: reputation.id,
        reputationTotalSupply: parseInt(daoState.reputationTotalSupply),
        fundingGoal: fundingGoal.toString(),
        minFeeToJoin: minFeeToJoin.toString(),
        memberReputation: memberReputation.toString()
      }
      await db.collection('daos').doc(dao.id).set(doc)
      const msg =`Updated dao ${dao.id}`
      response.push(msg)
      console.log(msg)
    } catch(err) {
      console.log(err)
      throw err
    }
    }
  }
  return response.join('\n')
}

async function updatePlugins() {

}

async function updateProposals() {
  arc.proposals({}, {subscribe: true, fetchAllData: true})
    .subscribe(async proposals => {
      proposals.map(async proposal => {
        const {coreState} = proposal;
        const proposalObject = {

          boostedAt: coreState.boostedAt,
          createdAt: coreState.createdAt,
          dao: coreState.dao.id,
          description: coreState.description,
          expiresInQueueAt: coreState.expiresInQueueAt,
          executionState: coreState.executionState,
          executed: coreState.executed,
          funding: coreState.funding.toString(),
          executedAt: coreState.executedAt,
          joinAndQuit: {
            funding: coreState.funding.toString(),
            proposedMemberId: coreState.proposer,
            proposedMemberAddress: coreState.proposer,
          },
          preBoostedAt: coreState.preBoostedAt,
          id: coreState.id,
          name: coreState.name,
          proposer: coreState.proposer,
          proposerId: coreState.proposer,
          resolvedAt: coreState.resolvedAt,
          stage: coreState.stage,
          title: coreState.title,
          type: coreState.type,
          votes: {
            list: coreState.votes,
            votesAgainst: parseInt(coreState.votesAgainst),
            votesFor: parseInt(coreState.votesFor),
          },
          links: [{
            title: "website",
            url: coreState.url,
          }],
          images: [],
          winningOutcome: coreState.winningOutcome,
        }
        await db.collection('proposals').doc(coreState.id).set(proposalObject)
      });

    })
}

// function updateDaosSubscription() {
//   //loop that runs a function every 15 seconds for 3 intervals
//   const response = ['xxx']
//   try {
//     const db = admin.firestore();
//     arc
//       .daos({}, {subscribe: true, fetchAllData: true})
//       .subscribe(res => {
//         res.map(async (dao, i) => {
//           // const state = await dao.fetchState()
//           // console.log(state)
//           const {id, address ,
//             memberCount,
//             name,
//             numberOfBoostedProposals ,
//             numberOfPreBoostedProposals ,
//             numberOfQueuedProposals ,
//             register,
//             reputation,
//             reputationTotalSupply,
//           } = dao.coreState;
//           // if (!dao.coreState.name.includes('Test DAO') && !dao.coreState.name.includes('Car DAO')) {
//           if (true) {
//             const joinAndQuitPlugins = await dao.plugins({where: {name: 'JoinAndQuit'}}).first()
//             if (joinAndQuitPlugins.length == 0) {
//               // we'll delete this DAO, as it does not have the correct plugins
//               try {
//                 console.log('DELETING: ', id)
//                 response.append(`Deleted dao ${id}`)
//                 await db.collection('daos').doc(id).delete();
//               } catch (e) {
//                 console.error('Failed to updated DAOs: ', error);
//               }
//             } else {
//               console.log('JaQ PLUGINS: ', joinAndQuitPlugins);
//               const joinAndQuitPlugin = joinAndQuitPlugins[0]
//               const {
//                 fundingGoal,
//                 minFeeToJoin,
//                 memberReputation
//               } = joinAndQuitPlugin.coreState.pluginParams;
//               try {
//                 await db.collection('daos').doc(id).set({
//                   id,
//                   address,
//                   memberCount,
//                   name,
//                   numberOfBoostedProposals,
//                   numberOfPreBoostedProposals,
//                   numberOfQueuedProposals,
//                   register,
//                   reputationId: reputation.id,
//                   reputationTotalSupply: parseInt(reputationTotalSupply),
//                   fundingGoal: fundingGoal.toString(),
//                   minFeeToJoin: minFeeToJoin.toString(),
//                   memberReputation: memberReputation.toString()
//                 })
//                 console.log(`[ Updated DAO ${name}@${address}] `);
//                 response.append(`updated DAO ${name}@${id}`)
//               } catch (error) {
//                 console.error(msg)
//                 const msg = `Failed to updated DAOs: ${error}`;
//                 throw Error(msg)
//               }
//             }
//           }
//           // if (dao.coreState.name.includes('Test DAO') || dao.coreState.name.includes('Car DAO')) {
//           //   console.log('DELETING: ', id)
//           //   await db.collection('daos').doc(id).delete();
//           // }
//         })
//       });
//   } catch(e) {
//     console.log('Error querying DAOs: ', e)
//     throw(e)
//   }
//   return '\n'.join(response)
// }

module.exports = {
  updateDaos,
  updatePlugins,
  updateProposals
}
