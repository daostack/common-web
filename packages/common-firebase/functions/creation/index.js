const functions = require('firebase-functions');
const express = require('express');
const cors = require('cors');
const bodyParser = require('body-parser');
const { responseExecutor, responseCreateExecutor } = require('../util/responseExecutor');
const { createCommon, createCommonTransaction } = require('./Common')
const { createRequestToJoinTransaction, createRequestToJoin  } = require('./RequestToJoin')
const { createVoteProposalTransaction, voteProposal  } = require('./VoteProposal')
const { createFundingProposalTransaction, createFundingProposal } = require('./FundingProposal');

const runtimeOptions = {
  timeoutSeconds: 540, // Maximum time 9 mins
}

const create = express();
create.use(bodyParser.json());
create.use(bodyParser.urlencoded({
  extended: true
}));
create.use(express.json());
create.use(express.urlencoded({ extended: true }));
create.use(cors({ origin: true }));

create.post('/createCommonTransaction', async (req, res) => {
  responseCreateExecutor(
    async () => {
      return await createCommonTransaction(req)
    },
    {
      req,
      res,
      successMessage: `Create common transaction successfully!`,
      errorMessage: `Unable to create a common transaction`
    }
  );
})

create.post('/createCommon', async (req, res) => {
  responseExecutor(
    async () => {
      return await createCommon(req);
    },
    {
      req,
      res,
      successMessage: `Common Created successfully!`,
      errorMessage: `Unable to create common!`
    }
  );
})

create.post('/createRequestToJoinTransaction', async (req, res) => {
  responseCreateExecutor(
    async () => {
      return await createRequestToJoinTransaction(req);
    },
    {
      req,
      res,
      successMessage: `Create requestToJoin transaction successfully!`,
      errorMessage: `Unable to create a request to join transaction!`
    }
  );
})

create.post('/createRequestToJoin', async (req, res) => {
  responseExecutor(
    async () => {
      return await createRequestToJoin(req);
    },
    {
      req,
      res,
      successMessage: `Request to join successfully!`,
      errorMessage: `Unable to create a request to join!`
    }
  );
})

  create.post('/createFundingProposalTransaction', async (req, res) => {
    responseCreateExecutor(
      async () => {
        return await createFundingProposalTransaction(req);
      },
      {
        req,
        res,
        successMessage: `Create funding proposal transaction successfully!!`,
        errorMessage: `Unable to create a funding proposal transaction`
      }
    );
})
  
  create.post('/createFundingProposal', async (req, res) => {
    responseExecutor(
      async () => {
        return await createFundingProposal(req);
      },
      {
        req,
        res,
        successMessage: `Create funding proposal successfully!`,
        errorMessage: `Unable to create a funding proposal !`
      }
    );
})

  create.post('/createVoteProposalTransaction', async (req, res) => {
    responseCreateExecutor(
      async () => {
        return await createVoteProposalTransaction(req);
      },
      {
        req,
        res,
        successMessage: `Create a vote proposal transaction successfully!`,
        errorMessage: `Unable to create vote transaction!`
      }
    );
  })
  
  create.post('/votePropoal', async (req, res) => {
    responseExecutor(
      async () => {
        return await voteProposal(req);
      },
      {
        req,
        res,
        successMessage: `Vote for the proposal successfully!`,
        errorMessage: `Unable to vote for a proposal !`
      }
    );
  })

exports.create = functions.runWith(runtimeOptions).https.onRequest(create);