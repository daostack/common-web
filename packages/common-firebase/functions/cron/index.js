const { backup } = require('./backupCron');

exports.crons = {
  backup
};