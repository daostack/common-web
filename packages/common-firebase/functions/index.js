require('module-alias/register');

const admin = require('firebase-admin');
const { databaseURL } = require('./settings');
const {env} = require('@env');

if(env.environment === 'dev') {
  admin.initializeApp();
} else {
  admin.initializeApp({
    credential: admin.credential.cert(require('@env/adminsdk-keys.json')),
    databaseURL: databaseURL
  });
}

const relayer = require('./relayer');
const graphql = require('./graphql');
const graphqlTriggers = require('./graphql/util/triggers');
const mangopay = require('./mangopay');
const mangopayTriggers = require('./mangopay/triggers');

// Add the __tests__ endpoints only if enabled
if(env.tests.enabled) {
  exports.tests = require('./tests').tests;
}

exports.relayer = relayer.relayer;
exports.graphql = graphql.graphql;
exports.mangopay = mangopay.mangopay;
exports.mangopaySubs = mangopayTriggers;
exports.graphqlSubs = graphqlTriggers;
// Disable notification
// exports.notification = require('./notification');

exports.cronJobs = require('./cron').crons;
