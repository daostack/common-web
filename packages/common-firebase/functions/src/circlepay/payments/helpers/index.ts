export { isFinalized, isSuccessful } from './statusHelper';

export { updatePayments } from './converter';

export { feesHelper } from './feesHelper';
export { failureHelper } from './failureHelper';