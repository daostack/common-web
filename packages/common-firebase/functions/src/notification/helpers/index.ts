export const getFundingRequestAcceptedTemplate = (country: string, amount: number) => {
  if (amount) {
    return !country
      ? 'userFundingRequestAcceptedUnknown'
      : (country === 'IL'
        ? 'userFundingRequestAcceptedIsraeli'
        : 'userFundingRequestAcceptedForeign');
  }
  return 'userFundingRequestAcceptedZeroAmount';
};