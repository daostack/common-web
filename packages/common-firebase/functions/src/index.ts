import { env } from './env';
import { tests } from './tests';
import * as cron from './cron';
import * as relayer from './relayer';
import * as graphql from './graphql';
import graphqlTriggers from './graphql/util/triggers';
import * as notification from './notification';
import * as event from './event';
import * as create from './creation';
import * as circlepay from './circlepay';
import * as circlepayTriggers from './circlepay/triggers';
import * as emailTriggers from './email/triggers';

// Add the __tests__ endpoints only if enabled
if(env.tests.enabled) {
  exports.tests = tests;
}

exports.relayer = relayer.relayer;
exports.graphql = graphql.graphql;
exports.graphqlSubs = graphqlTriggers;
exports.create = create.create;
exports.notificationSub = notification;
exports.eventSub = event;
exports.circlepay = circlepay.circlepay;
exports.circlepayTriggers = circlepayTriggers;
exports.emailTriggers = emailTriggers

exports.cronJobs = cron.crons;
