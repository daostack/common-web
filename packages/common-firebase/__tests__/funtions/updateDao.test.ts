import supertest from 'supertest';

import { runTest } from '../helpers/runTest';

import "../../functions/src";


runTest((funcs) => {
  const graphql = supertest(funcs.graphql);

  describe('Update dao by id ', () => {
    it('should fail if no id is provided', async () => {
      const res = await graphql.get('/update-dao-by-id');

      expect(res.ok).toBeFalsy();
      expect(res.status).toBe(500);

      expect(res.body.error).not.toBe(null);
      expect(res.body).toMatchSnapshot();
    });
  });
});
