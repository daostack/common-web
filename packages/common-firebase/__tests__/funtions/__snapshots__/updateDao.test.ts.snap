// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Functions Update dao by id  should fail if no id is provided 1`] = `
Object {
  "error": "You must provide a daoId (current value is \\"undefined\\") (7d5a61eb-a41d-47f9-b546-10840c0f3a02)",
  "errorCode": "GenericError",
  "errorId": "7d5a61eb-a41d-47f9-b546-10840c0f3a02",
  "errorMessage": "Something bad happened",
  "request": Object {
    "body": Object {},
    "headers": Object {
      "accept-encoding": "gzip, deflate",
      "connection": "close",
      "host": "127.0.0.1:62103",
      "user-agent": "node-superagent/3.8.3",
    },
    "query": Object {},
  },
}
`;
